import pytest
import shutil
from pathlib import Path
import pypipegraph as ppg
import mbf_align
import pysam
from mbf_qualitycontrol.testing import assert_image_equal
from mbf_sampledata import get_sample_data, get_sample_path
from mbf_qualitycontrol import prune_qc, get_qc_jobs


@pytest.mark.usefixtures("new_pipegraph_no_qc")
class TestAligned:
    def test_from_existing_bam(self):
        bam_path = get_sample_data(Path("mbf_align/ex2.bam"))
        bam_job = ppg.FileInvariant(bam_path)
        genome = object()
        lane = mbf_align.AlignedSample("test_lane", bam_job, genome, False, "AA123")
        assert lane.name == "test_lane"
        assert lane.load()[0] is bam_job
        assert isinstance(lane.load()[1], ppg.FileInvariant)
        assert lane.genome is genome
        assert not lane.is_paired
        assert lane.vid == "AA123"

        with pytest.raises(ValueError):
            mbf_align.AlignedSample("test_lane", bam_job, genome, False, "AA123")
        lane2 = mbf_align.AlignedSample("test_lane2", bam_job, genome, True, "AA123")
        assert lane2.is_paired

        b = lane.get_bam()
        assert isinstance(b, pysam.Samfile)
        b = lane.get_unique_aligned_bam()
        assert isinstance(b, pysam.Samfile)
        assert lane.get_bam_names()[0] == bam_path
        assert lane.get_bam_names()[1] == bam_path + ".bai"

        assert lane.mapped_reads() == 8
        assert lane.unmapped_reads() == 0
        for job in get_qc_jobs():
            assert job._pruned

    def test_lane_invariants_on_non_accepted_value(self):
        genome = object()
        with pytest.raises(ValueError):
            mbf_align.AlignedSample("test_lane", 123, genome, False, "AA123")

    def test_lane_raises_on_multifilegeneratingJobWithTwoBams(self):
        mfg = ppg.MultiFileGeneratingJob(["a.bam", "b.bam"], lambda: 5)
        genome = object()
        with pytest.raises(ValueError):
            mbf_align.AlignedSample("test_lane", mfg, genome, False, "AA123")

    def test_lane_raises_on_multifilegeneratingJobWithTwoBais(self):
        mfg = ppg.MultiFileGeneratingJob(["a.bam", "a.bam.bai", "b.bam.bai"], lambda: 5)
        genome = object()
        with pytest.raises(ValueError):
            mbf_align.AlignedSample("test_lane", mfg, genome, False, "AA123")

    def test_lane_raises_on_multifilegeneratingJobWithNoBAM(self):
        mfg = ppg.MultiFileGeneratingJob(["a.sam"], lambda: 5)
        genome = object()
        with pytest.raises(ValueError):
            mbf_align.AlignedSample("test_lane", mfg, genome, False, "AA123")

    def test_lane_invariants_on_string(self):
        bam_path = get_sample_data(Path("mbf_align/ex2.bam"))
        genome = object()
        lane = mbf_align.AlignedSample("test_lane", bam_path, genome, False, "AA123")
        assert isinstance(lane.load()[0], ppg.FileInvariant)

    def test_missing_index_file(self):
        bam_path = get_sample_data(Path("mbf_align/ex2.bam"))
        no_index = "noindex.bam"
        shutil.copy(bam_path, no_index)
        genome = object()
        lane = mbf_align.AlignedSample("test_lane", no_index, genome, False, "AA123")
        assert isinstance(lane.load()[0], ppg.FileInvariant)
        assert isinstance(lane.load()[1], ppg.FileGeneratingJob)
        assert lane.load()[1].job_id != "noindex.bam.bai"
        assert lane.load()[0] in lane.load()[1].prerequisites
        with pytest.raises(FileNotFoundError):
            lane.mapped_reads()
        ppg.run_pipegraph()
        assert lane.mapped_reads() == 8

    def test_creating_index_for_fg_job(self):
        def gen():
            shutil.copy(get_sample_data(Path("mbf_align/ex2.bam")), "sample.bam")

        ppg.util.global_pipegraph.quiet = False

        job = ppg.FileGeneratingJob("sample.bam", gen)
        genome = object()
        lane = mbf_align.AlignedSample("test_lane", job, genome, False, "AA123")
        assert isinstance(lane.load()[1], ppg.FileGeneratingJob)
        assert lane.load()[0] in lane.load()[1].prerequisites
        ppg.run_pipegraph()
        assert Path("sample.bam").exists()
        assert Path("sample.bam.bai").exists()

    def test_creating_index_for_mfg(self):
        def gen():
            shutil.copy(get_sample_data(Path("mbf_align/ex2.bam")), "sample.bam")

        ppg.util.global_pipegraph.quiet = False

        job = ppg.MultiFileGeneratingJob(["sample.bam"], gen)
        genome = object()
        lane = mbf_align.AlignedSample("test_lane", job, genome, False, "AA123")
        assert isinstance(lane.load()[1], ppg.FileGeneratingJob)
        assert lane.load()[0] in lane.load()[1].prerequisites
        ppg.run_pipegraph()
        assert Path("sample.bam").exists()
        assert Path("sample.bam.bai").exists()

    def test_subtraction_by_read(self):
        from mbf_sampledata import get_human_22_fake_genome

        genome = get_human_22_fake_genome()
        lane = mbf_align.AlignedSample(
            "test_lane",
            get_sample_data(Path("mbf_align/rnaseq_spliced_chr22.bam")),
            genome,
            False,
            "AA123",
        )  # index creation is automatic
        lane2 = mbf_align.AlignedSample(
            "test_lane2",
            get_sample_data(Path("mbf_align/rnaseq_spliced_chr22.bam")),
            genome,
            False,
            "AA124",
        )  # index creation is automatic
        lane3 = mbf_align.AlignedSample(
            "test_lane3",
            get_sample_data(Path("mbf_align/chipseq_chr22.bam")),
            genome,
            False,
            "AA123",
        )  # index creation is automatic
        lane3_subset = mbf_align.AlignedSample(
            "test_lane3_subset",
            get_sample_data(Path("mbf_align/chipseq_chr22_subset.bam")),
            genome,
            False,
            "AA123",
        )  # index creation is automatic

        lane_empty = lane.post_process(
            mbf_align.post_process.SubtractOtherLane(lane2), new_name="empty"
        )
        lane_full = lane.post_process(
            mbf_align.post_process.SubtractOtherLane(lane3), new_name="full"
        )
        lane_some = lane3.post_process(
            mbf_align.post_process.SubtractOtherLane(lane3_subset),
            result_dir="results/aligned/shu",
        )
        qc_jobs = [lane_some.post_processor_qc_jobs, lane_full.post_processor_qc_jobs]
        prune_qc(lambda job: job in qc_jobs)
        ppg.run_pipegraph()
        assert Path(lane_empty.get_bam_names()[1]).exists()
        assert Path(lane_full.get_bam_names()[1]).exists()
        assert lane_empty.mapped_reads() == 0
        assert lane_full.mapped_reads() == lane.mapped_reads()
        assert lane.mapped_reads() != 0
        assert (
            lane_some.mapped_reads()
            == lane3.mapped_reads() - lane3_subset.mapped_reads()
        )
        assert lane3_subset.mapped_reads()  # make sure there was something to subtract
        assert "shu" in lane_some.get_bam_names()[0]
        assert_image_equal(qc_jobs[0].filenames[0], "_result_dir")
        assert_image_equal(qc_jobs[0].filenames[0])


@pytest.mark.usefixtures("new_pipegraph")
class TestQualityControl:
    def prep_lane(self):
        from mbf_sampledata import get_human_22_fake_genome

        # straight from chr22 of the human genome
        genome = get_human_22_fake_genome()

        lane = mbf_align.AlignedSample(
            "test_lane",
            get_sample_data(Path("mbf_align/rnaseq_spliced_chr22.bam")),
            genome,
            False,
            "AA123",
        )
        return lane

    def _test_qc_plots(self, filename, remaining_job_count, chdir="."):
        lane = self.prep_lane()
        prune_qc(lambda job: filename in job.job_id)
        not_pruned_count = sum([1 for x in get_qc_jobs() if not x._pruned])
        assert not_pruned_count == remaining_job_count  # plot cache, plot_table, plot
        ppg.run_pipegraph()
        assert_image_equal(
            lane.result_dir / chdir / f"{lane.name}_{filename}", suffix="_" + filename
        )

    def test_qc_complexity(self):
        self._test_qc_plots("complexity.png", 3)

    def test_qc_strandedness(self):
        self._test_qc_plots("strandedness.png", 3)

    def test_qc_reads_per_biotype(self):
        self._test_qc_plots("reads_per_biotype.png", 1)

    def test_qc_alignment_statistics(self):
        self._test_qc_plots("alignment_statistics.png", 1, "..")

    def test_qc_subchromal_distribution(self):
        self._test_qc_plots("subchromosomal_distribution.png", 3)

    def test_qc_splice_sites(self):
        self._test_qc_plots("splice_sites.png", 3)

    def test_alignment_stats(self):
        from mbf_sampledata import get_human_22_fake_genome

        genome = get_human_22_fake_genome()
        lane = mbf_align.AlignedSample(
            "test_lane",
            get_sample_data(Path("mbf_align/rnaseq_spliced_chr22.bam")),
            genome,
            False,
            "AA123",
        )  # index creation is automatic
        counts = {"get_bam": 0}

        def get_bam():
            counts["get_bam"] += 1

            class DummySam:
                mapped = 5
                unmapped = 10

                def __enter__(self):
                    return self

                def __exit__(self, *args):
                    pass

            return DummySam()

        lane.get_bam = get_bam
        assert lane.get_alignment_stats() == {"Mapped": 5, "Unmapped": 10}
        assert counts["get_bam"] == 1

        class DummyAlignerWithout:
            pass

        lane = mbf_align.AlignedSample(
            "test_lane2",
            get_sample_data(Path("mbf_align/rnaseq_spliced_chr22.bam")),
            genome,
            False,
            "AA123",
            aligner=DummyAlignerWithout(),
        )  # index creation is automatic
        lane.get_bam = get_bam
        assert counts["get_bam"] == 1
        assert lane.get_alignment_stats() == {"Mapped": 5, "Unmapped": 10}
        assert counts["get_bam"] == 2

        class DummyAlignerWith:
            def get_alignment_stats(self, bam_filename):
                assert (
                    Path(bam_filename).resolve()
                    == get_sample_path("mbf_align/rnaseq_spliced_chr22.bam").resolve()
                )
                return {"Hello": 23}

        lane = mbf_align.AlignedSample(
            "test_lane3",
            get_sample_data("mbf_align/rnaseq_spliced_chr22.bam"),
            genome,
            False,
            "AA123",
            aligner=DummyAlignerWith(),
        )  # index creation is automatic
        lane.get_bam = get_bam
        assert counts["get_bam"] == 2
        assert lane.get_alignment_stats() == {"Hello": 23}
        assert counts["get_bam"] == 2
